Reference docs for experimental `captureOwnerStack` (#7427)
reactjs/react.dev: src/content/reference/react/captureOwnerStack.md

---
title: captureOwnerStack
---

<Wip>

**This API is experimental and is not available in a stable version of React yet.**

You can try it by upgrading React packages to the most recent experimental version:

- `react@experimental`

Experimental versions of React may contain bugs. Don't use them in production.

</Wip>

<Intro>

> `captureOwnerStack` reads the current **owner** Component stack and returns it as a string if available.

Reviewer: We should make it more clear that this is dev-only. I would mention it here, in a Caveat, and a Troubleshooting "My owner stacks are empty in production" section.

Author: Did all that. Also explained this in the createRoot and hydrateRoot usage examples. Though possible I missed some spots.

```js

> captureOwnerStack();

Reviewer: ```suggestion
const stack = captureOwnerStack();
```

```

</Intro>

<InlineToc />

---

## Reference {/*reference*/}

### `captureOwnerStack()` {/*captureownerstack*/}

Call `captureOwnerStack` to get the current owner Component stack.

```js
import * as React from 'react';

function Component() {
  if (process.env.NODE_ENV !== 'production') {
    const ownerStack = React.captureOwnerStack();
    console.log(ownerStack);
  }
}
```

#### Parameters {/*parameters*/}

`captureOwnerStack` does not take any parameters.

#### Returns {/*returns*/}

`captureOwnerStack` returns `string | null`.

If no owner stack is available, it returns an empty string.
Outside of development builds, `null` is returned.

> ## Owner Component stacks vs parent Component stacks {/*owner-component-stacks-vs-parent-component-stacks*/}

Reviewer: Let's wrap this in a `<DeepDive />`

Author: Done

> The owner Component stack is different from the **parent** Component stack available error handlers like [`errorInfo.componentStack` in `onUncaughtError`](http://localhost:3000/reference/react-dom/client/hydrateRoot#show-a-dialog-for-uncaught-errors).

Reviewer: localhost link

      <legend>{label}</legend>
      <SubComponent key={label} disabled={label === 'disabled'} />
    </fieldset>
  );
}

function Navigation() {
  return null;
}

function App({children}) {
  return (
    <React.Suspense fallback="loading...">
      <main>
        <Navigation />
        {children}
      </main>
    </React.Suspense>
  );
}

createRoot(document.createElement('div')).render(
  <App>
    <Component label="disabled" />
  </App>
);
```

`SubComponent` would throw an error.
The **parent** component stack of that error would be

```
at SubComponent
at fieldset
at Component
at main
at React.Suspense
at App
```

However, the owner stack would only read

```
at SubComponent
at Component
```

Neither `App` nor the DOM components (e.g. `fieldset`) are considered owners in this stack since they didn't contribute to "creating" the node containing `SubComponent`. `App` and DOM components only forwarded the node. `App` just rendered the `children` node as opposed to `Component` which created a node containing `SubComponent` via `<SubComponent />`.

Neither `Navigation` nor `legend` are in the stack at all since it's only a sibling to a node containing `<SubComponent />`.

> ### When to use which {/*when-to-use-which*/}

Reviewer: This explanation seems too technical to really understand when to use each. This is more "what they represent" and not "when to use". It's not really an either/or either. Owner stacks can be added in addition to component stacks when you're in DEV and want to add additional debugging info.

Author: >  It's not really an either/or either.

I agree. This is not what I wanted to convey with this section.

Author: Removed the section. It didn't really say anything new we already explained earlier.

The owner stack is useful for tracing the flow of props. Owner stacks are equivalent to [call stacks](https://developer.mozilla.org/en-US/docs/Glossary/Call_stack) if all JSX would be converted to direct function calls e.g. `Component({label: "disabled"})` instead of `<Component label="disabled" />`.

## Usage {/*usage*/}

### Expanding error stacks {/*expanding-error-stacks*/}

In addition to the stack trace of the <CodeStep step={1}>error</CodeStep> itself, you can use <CodeStep step={2}>`captureOwnerStack`</CodeStep> to append the stack trace of the owner Component.
This can help trace the error especially when the error is caused by props. The owner Component stack helps trace the flow of props.

```jsx [[9, 15, "error"], [34, 10, "captureOwnerStack"]]
import { captureOwnerStack } from 'react'
import { hydrateRoot } from 'react-dom/client';

const root = hydrateRoot(
  document.getElementById('root'),
  <App />,
  {
    onCaughtError: (error, errorInfo) => {
      if (process.env.NODE_ENV !== 'production') {
        const ownerStack = captureOwnerStack();

>         error.stack += ownerStack;

Reviewer: Do we really recommend appending it to the existing stack? Won't that add it to the end?

IMO this should be like:

```
if (__DEV__) { 
  showDialog(error, ownerStack, componentStack);
} else {
  logProdError(error, componentStack);
}
```

Author: We do that in Next.js and it's conceptually how you arrive at the error if every JSX callsite would just be a function call.

---
title: captureOwnerStack
---

<Wip>

> **This API is experimental and is not available in a stable version of React yet.**

Reviewer: I think we should update this to the Canary badge, and just wait until we add this to Canary to merge the docs.

Author: Makes sense. Need to remember to update the sandboxes to use the Canary channel. I'll leave them on experimental for now so that we can verify they're working.

You can try it by upgrading React packages to the most recent experimental version:

- `react@experimental`

Experimental versions of React may contain bugs. Don't use them in production.

</Wip>

<Intro>

`captureOwnerStack` reads the current Owner Stack in development and returns it as a string if available.

```js
const stack = captureOwnerStack();
```

</Intro>

<InlineToc />

---

## Reference {/*reference*/}

### `captureOwnerStack()` {/*captureownerstack*/}

Call `captureOwnerStack` to get the current Owner Stack.

```js {5,5}
import {captureOwnerStack} from 'react';

function Component() {
  if (process.env.NODE_ENV !== 'production') {
    const ownerStack = captureOwnerStack();
    console.log(ownerStack);
  }
}
```

#### Parameters {/*parameters*/}

`captureOwnerStack` does not take any parameters.

#### Returns {/*returns*/}

`captureOwnerStack` returns `string | null`.

> If no Owner Stack is available, it returns an empty string.
> Outside of development builds, `null` is returned.

Reviewer: ```suggestion
If no Owner Stack is available, it returns an empty string. Outside of development builds, `null` is returned.
```

Also, instead of "if no owner stack is available" we should say when owner stacks are not available. Like maybe not too specific, but something like "if it's not called inside a component lifecycle".

Reviewer: We call this out in the troubleshooting section so we should just also include that explanation here.

Author: component lifecycle can read too specific. It's not obvious that `onCaughtError` is part of the component lifecycle. 

I'll be more specific here though.

For example, consider the following code:

<Sandpack>

```js src/App.js
import {Suspense} from 'react';

function SubComponent({disabled}) {
  if (disabled) {
    throw new Error('disabled');
  }
}

export function Component({label}) {
  return (
    <fieldset>
      <legend>{label}</legend>
      <SubComponent key={label} disabled={label === 'disabled'} />
    </fieldset>
  );
}

function Navigation() {
  return null;
}

export default function App({children}) {
  return (
    <Suspense fallback="loading...">
      <main>
        <Navigation />
        {children}
      </main>
    </Suspense>
  );
}
```

```js src/index.js
import {captureOwnerStack} from 'react';
import {createRoot} from 'react-dom/client';
import App, {Component} from './App.js';
import './styles.css';

createRoot(document.createElement('div'), {
  onUncaughtError: (error, errorInfo) => {
    // The stacks are logged instead of showing them in the UI directly to highlight that browsers will apply sourcemaps to the logged stacks.

>     // Note that sourcemapping is only applied in the real browser console not in the fake one displayed on this page.

Reviewer: Would shorten the line length of this comment so you don't force a hscroll on desktop at least:

<img width="868" alt="Screenshot 2025-01-19 at 10 42 53 AM" src="https://github.com/user-attachments/assets/5f2e467f-cd85-45a1-8cd7-2be4f0d6abd1" />

Author: done

>     console.log(errorInfo.componentStack);
>     console.log(captureOwnerStack());

Reviewer: Would add labels:

```suggestion
    console.log('Component Stack:', errorInfo.componentStack);
    console.log('Owner Stack:', captureOwnerStack());
```
<img width="888" alt="Screenshot 2025-01-19 at 10 46 52 AM" src="https://github.com/user-attachments/assets/7cbbc5cd-86b8-4c61-aea6-80ea779d2b9e" />

  },
}).render(
  <App>
    <Component label="disabled" />
  </App>
);
```

```json package.json hidden
{
  "dependencies": {
    "react": "experimental",
    "react-dom": "experimental",
    "react-scripts": "latest"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject"
  }
}
```

```html public/index.html hidden
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <p>Check the console output.</p>
  </body>
</html>
```

</Sandpack>

> `SubComponent` would throw an error.
> The Component Stack of that error would be

Reviewer: I've seen this in a few places - think there's way too many newlines in these docs, instead I would write it more paragraph style:

```suggestion
`SubComponent` would throw an error. The Component Stack of that error would be:
```

But you should also expand here, because it's not clear what SubComponent is:

```suggestion
In this example, `SubComponent` throws an error in render, which is not caught by an error boundary. The root `onUncaughtError` method is called, and logs the Component Stack and Owner Stack. The Component Stack of that error is:
```

Author: This i due to how we render markdown. Normally markdown doesn't render newlines if you put two sentences on two lines. I just got into the habit of it since it makes review and suggestions easier.

Will adjust.

Author: Should've caught all accidental newlines now.

```
at SubComponent
at fieldset
at Component
at main
at React.Suspense
at App
```

However, the Owner Stack would only read

```
at Component
```

Neither `App` nor the DOM components (e.g. `fieldset`) are considered Owners in this Stack since they didn't contribute to "creating" the node containing `SubComponent`. `App` and DOM components only forwarded the node. `App` just rendered the `children` node as opposed to `Component` which created a node containing `SubComponent` via `<SubComponent />`.

Neither `Navigation` nor `legend` are in the stack at all since it's only a sibling to a node containing `<SubComponent />`.

`SubComponent` is omitted because it's already part of the callstack.

</DeepDive>

## Usage {/*usage*/}

> ### Improve error reporting {/*improve-error-reporting*/}

Reviewer: When I read this and the usage below, I hear "reporting for production logging" which isn't the case. I really think the usages here are limited to just instrumenting DEV dialogs.

- [createRoot usage: Show a dialog for uncaught errors
](/reference/react-dom/client/createRoot#show-a-dialog-for-uncaught-errors)
- [createRoot usage: Displaying Error Boundary errors](/reference/react-dom/client/createRoot#show-a-dialog-for-uncaught-errors)
- [createRoot usage: Displaying a dialog for recoverable errors](/reference/react-dom/client/createRoot#displaying-a-dialog-for-recoverable-errors)

### Expanding error stacks {/*expanding-error-stacks*/}

In addition to the <CodeStep step={1}>stack trace of the error</CodeStep> itself, you can use <CodeStep step={2}>`captureOwnerStack`</CodeStep> to append the Owner Stack.
This can help trace the error especially when the error is caused by props. The Owner Stack helps trace the flow of props.

```js src/index.js [[1, 8, "error.stack"], [2, 7, "captureOwnerStack()"]]
import {captureOwnerStack} from 'react';
import {createRoot} from 'react-dom/client';

const root = createRoot(document.getElementById('root'), {
  onUncaughtError: (error, errorInfo) => {
    if (process.env.NODE_ENV !== 'production') {
      const ownerStack = captureOwnerStack();

>       error.stack += ownerStack;

Reviewer: Why would you do this? Isn't it already in the stack if you have DevTools open? And if DevTools is open, won't this double add the owner stack?

    }

    console.error('Uncaught', error);
  },
}).render(<App />);
```

<Sandpack>

```js
function useCustomHook() {
  throw new Error('Boom!');
}

function Component() {
  useCustomHook();
}

export default function App() {
  return <Component />;
}
```

```js src/index.js
import {captureOwnerStack} from 'react';
import {createRoot} from 'react-dom/client';
import App from './App.js';
import './styles.css';

const root = createRoot(document.getElementById('root'), {
  onUncaughtError: (error, errorInfo) => {
    if (process.env.NODE_ENV !== 'production') {
      const ownerStack = captureOwnerStack();
      error.stack =
        // The stack is only split because these sandboxes don't implement ignore-listing 3rd party frames via sourcemaps.
        // A framework would ignore-list stackframes from React via sourcemaps and then you could just `error.stack += ownerStack`.
        // To learn more about ignore-listing see https://developer.chrome.com/docs/devtools/x-google-ignore-list

>         error.stack.split('\n    at react-stack-bottom-frame')[0] + ownerStack;

Reviewer: This should be a separate usage section like "How to collapse frames in DEV dialogs without sourcemaps"

Reviewer: Or maybe Troubleshooting

Reviewer: This also seems weird to me because React DevTools adds the owner stack in the stack trace? Why would you want to do this?

Author: Removed this section. Better leave this to React Devtools instead of polyfills.

    }

    // The stacks are logged instead of showing them in the UI directly to highlight that browsers will apply sourcemaps to the logged stacks.
    // Note that sourcemapping is only applied in the real browser console not in the fake one displayed on this page.

>     console.error('Uncaught', error);

Reviewer: If you've patched `console.error` to add owner stacks, and you call console.error here, it will result in multiple owner stacks being applied. It's not clear to me whether you should only patch console.error, or if you should instrument the root handlers and console.error without them overlapping somehow. 

@sebmarkbage how should this work ideally? Maybe we're over-emphasizing the root handler case?

The way I expected this to work is like:

```
onUncaughtError() {
  if (process.env.NODE_ENV !== 'production') {
      const ownerStack = captureOwnerStack();
      sendDevOnlyLogs(error, ownerStack);
  } else {
    sendProductionLogs(error);
  }
  
  // Patch or DevTools will add logs to the console output.
  console.error(error);
}
```

Author: Yeah I think the usage in error handlers is the better example. If there's a need to document polyfilling what React Devtools is doing, we can add it later.

Reviewer: @eps1lon how do you implement a dev dialog without patching console.error? You can't only do the root handlers because you'll miss react warnings like key errors.

Author: This seems like a separate task and not strictly tied to `captureOwnerStack`. We haven't documented how to implement a dev dialog that displays `console.error` before so I don't see why this needs to be coupled with `captureOwnerStack`. 

In Next.js we're just reading errors from console. Owner stacks are then always displayed in the overlay via `captureOwnerStack` (no option). But displaying Component stacks directly in the console is left to React DevTools.

Reviewer: > We haven't documented how to implement a dev dialog that displays console.error

Yeah but we have for the other root handlers, and if you did that plus your own console.error patch before, then owner stacks break that implementation so you need docs for how to add it back. So we need to at least tell people how to get back to what worked before owner stacks.

  },
}).render(<App />);
```

```json package.json hidden
{
  "dependencies": {
    "react": "experimental",
    "react-dom": "experimental",
    "react-scripts": "latest"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject"
  }
}
```

```html public/index.html hidden
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <p>Check the console output.</p>
    <div id="root"></div>
  </body>
</html>
```

</Sandpack>

## Troubleshooting {/*troubleshooting*/}

### The Owner Stack is `null` {/*the-owner-stack-is-null*/}

> `captureOwnerStack` was called outside of development builds.
> For performance reasons, React will not keep track of Owners in production.

Reviewer: ```suggestion
`captureOwnerStack` was called outside of development builds. For performance reasons, React will not keep track of Owner Stacks in production.
```

Author: Done

### The Owner Stack is empty {/*the-owner-stack-is-empty*/}

> The call of `captureOwnerStack` happened outside of a React controlled function e.g. in a `setTimeout` callback.
> During render, Effects, Events, and React error handlers (e.g. `hydrateRoot#options.onCaughtError`) Owner Stacks should be available.

Reviewer: ```suggestion
The call of `captureOwnerStack` happened outside of a React controlled function e.g. in a `setTimeout` callback. During render, Effects, Events, and React error handlers (e.g. `hydrateRoot#options.onCaughtError`) Owner Stacks should be available.
```

And can we add a sandbox with the issue and fix?

Author: Added one for custom DOM event handlers which is also a likely pitfall.
